src/ngraph/ngraph_subgraph.cc

```diff
@@ -0,0 +1,231 @@
+/*******************************************************************************
+* Copyright 2018 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include <mxnet/ndarray.h>
+#include "../operator/subgraph/common.h"
+#include "../operator/subgraph/default_subgraph_op.h"
+#include "ngraph_nnvm_ops.h"
+
+namespace ngraph_bridge {
+using namespace nnvm;
+using namespace mxnet;
+using namespace mxnet::op;
+
+std::shared_ptr<ngraph_bridge::Graph> get_ngraph(const NodeAttrs& attrs) {
+  auto& g = nnvm::get<ngraph_bridge::NGraphParam>(attrs.parsed).g;
+  return g;
+}
+
+class NgraphSubgraphOperator {
+ public:
+  explicit NgraphSubgraphOperator(
+      std::shared_ptr<ngraph_bridge::Graph> ngraph) {
+    ngraph_ = ngraph;
+  }
+
+  void Forward(const OpContext& ctx, const std::vector<NDArray>& inputs,
+               const std::vector<OpReqType>& req,
+               const std::vector<NDArray>& outputs);
+  void Backward(const OpContext& ctx, const std::vector<NDArray>& inputs,
+                const std::vector<OpReqType>& req,
+                const std::vector<NDArray>& outputs);
+
+ private:
+  std::shared_ptr<ngraph_bridge::Graph> ngraph_;
+};
+
+void NgraphSubgraphOperator::Forward(const OpContext& ctx,
+                                     const std::vector<NDArray>& inputs,
+                                     const std::vector<OpReqType>& req,
+                                     const std::vector<NDArray>& outputs) {
+  compute_forward(ctx, ngraph_, inputs, req, outputs);
+}
+
+void NgraphSubgraphOperator::Backward(const OpContext& ctx,
+                                      const std::vector<NDArray>& inputs,
+                                      const std::vector<OpReqType>& req,
+                                      const std::vector<NDArray>& outputs) {
+  compute_backward(ctx, ngraph_, inputs, req, outputs);
+}
+
+OpStatePtr CreateNgraphSubgraphOpState(const NodeAttrs& attrs, Context ctx,
+                                       const std::vector<TShape>& in_shapes,
+                                       const std::vector<int>& in_types) {
+  return OpStatePtr::Create<NgraphSubgraphOperator>(get_ngraph(attrs));
+}
+
+OpStatePtr CreateNgraphBackwardOpState(const NodeAttrs& attrs, Context ctx,
```

Comment:
Do we need separate forward and backward op state functions?

Reply:
i think backopstate creation may need to be separate when we use subgraph fully (as the node attrs of forward would not have ngraph yet, but backward might). currently both reuse same ngraph attrs.parsed.